warehouse/accounts/services.py

```diff
@@ -759,3 +772,163 @@ def check_password(self, password, *, tags=None):
         # This service allows *every* password as a non-breached password. It will never
         # tell a user their password isn't good enough.
         return False
+
+
+class GitHubPublicKeyMetaAPIError(InvalidTokenLeakRequest):
+    pass
+
+
+@implementer(IGitHubTokenScanningPayloadVerifyService)
+class GitHubTokenScanningPayloadVerifyService:
+    """
+    Checks payload signature using:
+    - `requests` for HTTP calls
+    - `cryptography` for signature verification
+    """
+
+    def __init__(self, *, session, metrics, api_token=None):
+        self._metrics = metrics
+        self._session = session
+        self._api_token = api_token
+
+    @classmethod
+    def create_service(cls, context, request):
+        return cls(
+            session=request.http,
+            metrics=request.find_service(IMetricsService, context=context),
+            api_token=request.registry.settings["github.token"],
+        )
+
+    def verify(self, *, payload, key_id, signature):
+
+        try:
+            pubkey_api_data = self._retrieve_public_key_payload()
+            public_keys = self._extract_public_keys(pubkey_api_data)
+            public_key = self._check_public_key(
+                github_public_keys=public_keys, key_id=key_id
+            )
+            self._check_signature(
+                payload=payload, public_key=public_key, signature=signature
+            )
+        except InvalidTokenLeakRequest as exc:
+            self._metrics.increment(
+                f"warehouse.token_leak.github.auth.error.{exc.reason}"
+            )
+            return False
+
+        self._metrics.increment("warehouse.token_leak.github.auth.success")
+        return True
+
+    def _get_headers(self):
+        if self._api_token:
+            return {"Authorization": f"token {self._api_token}"}
+        return {}
+
+    def _retrieve_public_key_payload(self):
+        # TODO: cache ?
+
+        token_scanning_pubkey_api_url = (
+            "https://api.github.com/meta/public_keys/token_scanning"
+        )
+        headers = self._get_headers()
+        try:
+            response = self._session.get(token_scanning_pubkey_api_url, headers=headers)
+            response.raise_for_status()
+            return response.json()
+        except requests.HTTPError as exc:
+            # TODO Log, including status code and body
```

Comment:
This seems like an exception we would want to see and should just let occur.

Reply:
Yep ! I always like to `raise BlaError() from exc` just to make it super duper extra easy to figure out what happened, and have the code know this is semi-expected.